### PR TITLE
Fix type promotion for cosine_similarity()

### LINKED ISSUE #61454: torch.nn.functional.cosine_similarity got unexpected behavier with torch.int8 dtype
## 🐛 Bug

`torch.nn.functional.cosine_similarity` got unexpected behavier with torch.int8 dtype.

```
>>> a = torch.randn(10)*100
>>> a
tensor([ 47.8993, -27.2694,  20.9548, -13.5573, -54.0388, -76.8524,  -7.4037,
        -95.7477, 108.9276, -44.5625])
>>> b = a
>>> F.cosine_similarity(a, b, dim=-1)
tensor(1.)
>>> F.cosine_similarity(a.to(torch.int8), b, dim=-1)
tensor(3.5215e+12)
>>> F.cosine_similarity(a.to(torch.int16), b, dim=-1)
tensor(1.0000)
```

As we can see, `F.cosine_similarity(a.to(torch.int8), b, dim=-1)` got result of `tensor(3.5215e+12)`, while the expected result is `tensor(1.)`.

## Environment

PyTorch version: 1.9.0

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @albanD @mruberry @nairbv

### PR DESCRIPTION
**Pull request by jbschlosser**
Fixes #61454

### CODE REVIEW

**Review by albanD — APPROVED** [review]
LGTM!
Thanks for sending a fix for that!

**Review by nairbv — APPROVED** [review]
we might also want to update the doc for this function
